core/imt/src/RTaskArena.cxx

```diff
@@ -123,17 +123,14 @@ tbb::task_arena &RTaskArenaWrapper::Access()
    return *fTBBArena;
 }
 
-
 std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> GetGlobalTaskArena(unsigned maxConcurrency)
 {
    static std::weak_ptr<ROOT::Internal::RTaskArenaWrapper> weak_GTAWrapper;
-   if (weak_GTAWrapper.expired()) {
-      std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> shared_GTAWrapper(new ROOT::Internal::RTaskArenaWrapper(maxConcurrency));
-      weak_GTAWrapper = shared_GTAWrapper;
-      return weak_GTAWrapper.lock();
-   }
-
-   return weak_GTAWrapper.lock();
+   if (auto sp = weak_GTAWrapper.lock())
+      return sp;
+   auto sp = std::make_shared<ROOT::Internal::RTaskArenaWrapper>(maxConcurrency);
+   weak_GTAWrapper = sp;
+   return sp;
```

Comment:
I assume you can still have a race condition. If two threads try to get the task arena, and it doesn't exist, both will get a `nullptr`. Now, they both make a shared pointer to arenas. One arena will be discarded once the returned pointer goes out of scope.

Reply:
You are right! And even further, two threads could concurrently assign to `weak_GTAWrapper` which is also a race condition.

Now I ask myself what the semantic of this function should be? It currently seems like you want to destroy the task arena, once the last user is done with it. Why is this even needed? Why can be not just keep it around?

Like:
```
ROOT::Internal::RTaskArenaWrapper& GetGlobalTaskArena(unsigned maxConcurrency) {
  static ROOT::Internal::RTaskArenaWrapper taw{maxConcurrency};
  return taw;
}
```

I furthermore ask myself what the meaning of `maxConcurrency` is for the caller. Because depending on whether there is any other thread still keeping the task arena alive with a different value of `maxConcurrency` I ether get a task arena with the `maxConcurrency` I requested or a different one.

Reply:
True. So...you guys want me to put locks in there?

Reply:
No. Not now. First please clarify what are the semantics of this function and the lifetime of the task arena.

Reply:
About the race condition: iiuc you want to solve [ROOT-10762](https://sft.its.cern.ch/jira/browse/ROOT-10762) here, which I think is more of a nice-to-have than a must

Reply:
A function-local static would be thread safe, and live till the end of ROOT. Sounds ok to me.

Reply:
@eguiraud I do not want to solve any already existing issues. Just not create new ones ;)
@hageboeck If having a single task arena for the entire lifetime of ROOT is not wrong, than I guess the function local static is a safe bet.

Reply:
if i'm not mistaken, originally, the TPoolManager needed to be destructed and reconstructed to change the number of threads in the pool, so it could not have static lifetime. Xavi, am I right? If yes, could the new logic accommodate of working threads in a different way?

Reply:
Regarding the concurrency, I think we agreed that you can choose it once, and then it's fixed. We would need a mechanism to tell users, though, that if they input another value, it's getting ignored. E.g.
```
EnableIMT(2);
// do stuff
DisableIMT();
EnableIMT(4);
```

Reply:
this is not the current behavior and i don't remember agreeing it should be..?

Reply:
@bernhardmgruber Oh, but this issue is there already, currently in TPoolManager.

A single task arena for the entire lifetime of ROOT is **not** what we want. And the only way I see to solve the issue right now is to have a critical section.

The idea of this function is to have either none or a single global task_arena alive at any point in time within ROOT.

Right now, this function is called in two cases: When we Initialize a TThreadExecutor and when we call EnableImplicitMT. Both of them hold a shared ptr to the RTaskArenaWrapper during their lifetime, and the RTaskArenaWrapper should be destroyed when every shared owner ceases to exist (TThreadExecutor is destroyed) or releases it (DisableImplicitMT() releases the shared_ptr acquired by EnableImplicitMT). This allows for subsequent reinstantiation of the task_arena, potentially with a different number of worker threads.

Reply:
> Regarding the concurrency, I think we agreed that you can choose it once, and then it's fixed. We would need a mechanism to tell users, though, that if they input another value, it's getting ignored. E.g.
> 
> ```
> EnableIMT(2);
> // do stuff
> DisableIMT();
> EnableIMT(4);
> ```

as @eguiraud said, that is not the agreed current behaviour. But the mechanism to warn users when we ignore values is in the PR (and currently in TThreadExecutor but not IMT in master)

Reply:
@xvallspl thx for the clarification. Then the task arena has shared ownership/lifetime and a shared_ptr might be a good choice here. And as you pointed out, you might need a mutex here to protect the creation of a new task arena:
```
std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> GetGlobalTaskArena(unsigned maxConcurrency)
{
  static std::weak_ptr<ROOT::Internal::RTaskArenaWrapper> weak_GTAWrapper;
  if (auto sp = weak_GTAWrapper.lock())
    return sp;
  static std::mutex m;
  std::lock_guard _{m};
  if (auto sp = weak_GTAWrapper.lock())
    return sp;
  auto sp = std::make_shared<ROOT::Internal::RTaskArenaWrapper>(maxConcurrency);
  weak_GTAWrapper = sp;
  return sp;
}
```

Reply:
@eguiraud ok for this as well? nice-to have is enough. I don't expect anyone to start spamming enough TThreadExecutors concurrently for this to be a performance issue

Reply:
yep, creating TThreadExecutors in a parallel hot loop is not a usecase we care to optimize

Reply:
Looks to me like this will work.